L74-java-channel-creds.md

```diff
@@ -0,0 +1,571 @@
+`ChannelCredentials` will be introduced, which can be provided when
+constructing a `ManagedChannelBuilder':
+
+```java
+// instead of:
+channel = ManagedChannelBuilder.forTarget("example.com")
+    .build();
+channel = ManagedChannelBuilder.forTarget("example.com").usePlaintext()
+    .build();
+channel = ManagedChannelBuilder.forAddress("example.com", 443)
+    .build();
+// the user would now:
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilder("example.com", new InsecureChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilderForAddress("example.com", 443, new TlsChannelCredentials())
+    .build();
+```
+
+`newChannelBuilder()` will iterate through the `ManagedChannelProviders`
+trying to create a builder for each in turn. The first provider that succeeds
+will be used. If no provider is able to handle the credentials, then
+`newChannelBuilder()` will throw. `newChannelBuilderForAddress()` will be a
+convenience function that creates the target string and calls
+`newChannelBuilder()`, mirroring its current behavior.
+
+The new ChannelCredentials API cannot be mixed with the pre-existing
+`useTransportSecurity()` and `usePlaintext()` builder APIs. The old-style
+methods will throw if called on a builder with a ChannelCredential provided.
+
+Note that this new API does not allow: 1) creating a channel without a
+credential and supplying the credential later and 2) changing the credential
+after creating the builder. This is not expected to be much of an issue as
+users typically have the security information available before creating the
+builder and reusing a builder is rare, especially to change the security.
+
+```java
+package io.grpc;
+
+/**
+ * Represents a security configuration to be used for channels. There is no
+ * generic mechanism for processing arbitrary ChannelCredentials; the consumer
+ * of the credential (the channel) must support each implementation explicitly
+ * and separately. Consumers are not required to support all types or even all
+ * possible configurations for types that are partially supported, but they
+ * <em>must</em> at least fully support ChoiceChannelCredentials.
+ *
+ * <p>A ChannelCredential provides client identity and authenticates the server.
+ * This is different from CallCredentials, which only provides client identity.
+ * They can also influence types of encryption used and similar security
+ * configuration.
+ *
+ * The concrete credential type should not be relevant to most users of the API
+ * and may be an implementation decision. Users should generally use the {@code
+ * ChannelCredentials} type for variables instead of the concrete type.
+ * Freshly-constructed credentials should be returned as {@code
+ * ChannelCredentials} instead of a concrete type to encourage this pattern.
+ * Concrete types would only be used after {@code instanceof} checks (which must
+ * consider {@code ChoiceChannelCredentials}!).
```

Comment:
We are not hiding concrete types (like `ChoiceChannelCredentials` or `CompositeChannelCredentials`) but expect users to do `instanceof` and a type-cast to be able to use methods like `getChannelCredentials()` of `CompositeChannelCredentials` (for example). How does it support the view that "concrete type should not be relevant to API users and may be an implementation decision"?

Wouldn't it then be better for the `create()` methods to return the actual public concrete type, but making sure that most of the functionality is at the `ChannelCredentials` level and we only expose the minimum necessary functionality in the exposed concrete type?

Reply:
Most users don't need to call any methods on the credentials. They just 1) create them and 2) pass them around. The transports are the main thing that needs to call methods on the credentials.

> Wouldn't it then be better for the create() methods to return the actual public concrete type

Better in what way?

> but making sure that most of the functionality is at the ChannelCredentials level and we only expose the minimum necessary functionality in the exposed concrete type?

￼I don't understand what that would look like. Could you provide a sketch of what ChannelCredentials would look like in that approach?

Reply:
> > Wouldn't it then be better for the create() methods to return the actual public concrete type
> 
> Better in what way?

Compared to what was described in the original text "...Concrete types would only be used after {@code instanceof} checks..."

So with the current proposal one would do:
```
ChannelCredentials channelCreds = ChoiceChannelCredentials.create(...);
...
// somewhere later in the code
if (channelCreds instanceof ChoiceChannelCredentials) {
    ChoiceChannelCredentials choiceChannelCreds = (ChoiceChannelCredentials)channelCreds;
    // do something with choiceChannelCreds
}
```
Whereas if `create` returned the actual concrete type you wouldn't need the check and the cast:
```
ChoiceChannelCredentials choiceChannelCreds = ChoiceChannelCredentials.create(...);
...
// somewhere later in the code
// do something with choiceChannelCreds
```

All I am saying is that exposing/making public concrete subclasses but not have `create` return them looks a little inconsistent but I do understand the argument about encouraging the pattern.

> 
> > but making sure that most of the functionality is at the ChannelCredentials level and we only expose the minimum necessary functionality in the exposed concrete type?
> 
> ￼I don't understand what that would look like. Could you provide a sketch of what ChannelCredentials would look like in that approach?

Nothing different from what is currently being proposed. As you said users would mostly create creds and pass them around because the `ChannelCredentials` interface/functionality should work for most cases.

Reply:
There is a very strong producer/consumer split with this API. The producer _doesn't have any need_ to look at the configuration; it provided it. The consumer is what needs to look at the configuration, and if it only supports one specific credential type that sort of defeats the purpose. So I don't see your example as reflecting any actual use case for this API.

> As you said users would mostly create creds and pass them around because the ChannelCredentials interface/functionality should work for most cases.

No, I don't understand what configuration you think should be within ChannelCredentials.